src/Stratis.Bitcoin/Consensus/PHConsensusManagerBehavior.cs

```diff
@@ -0,0 +1,306 @@
+                }
+                catch (OperationCanceledException)
+                {
+                    this.logger.LogTrace("Unable to send headers message to peer '{0}'.", peer.RemoteSocketEndpoint);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Processes "headers" message received from the peer.
+        /// </summary>
+        /// <param name="peer">Peer from which the message was received.</param>
+        /// <param name="headersPayload">Payload of "headers" message to process.</param>
+        /// <remarks>
+        /// "headers" message is sent in response to "getheaders" message or it is solicited
+        /// by the peer when a new block is validated (unless in IBD).
+        /// <para>
+        /// When we receive "headers" message from the peer, we can adjust our knowledge
+        /// of the peer's view of the chain. We update its pending tip, which represents
+        /// the tip of the best chain we think the peer has.
+        /// </para>
+        /// </remarks>
+        protected async Task ProcessHeadersAsync(INetworkPeer peer, ProvenHeadersPayload headersPayload)
+        {
+            List<ProvenBlockHeader> headers = headersPayload.Headers;
+
+            if (headers.Count == 0)
+            {
+                this.logger.LogTrace("Proven Headers payload with no headers was received. Assuming we're synced with the peer.");
+                this.logger.LogTrace("(-)[NO_HEADERS]");
+                return;
+            }
+
+            if (!this.ValidateHeadersPayload(peer, headersPayload, out string validationError))
+            {
+                this.peerBanning.BanAndDisconnectPeer(peer.PeerEndPoint, validationError);
+
+                this.logger.LogTrace("(-)[VALIDATION_FAILED]");
+                return;
+            }
+
+            using (await this.asyncLock.LockAsync().ConfigureAwait(false))
+            {
+                if (this.cachedHeaders.Count > CacheSyncHeadersThreshold) // TODO when proven headers are implemented combine this with size threshold of N mb.
+                {
+                    // Ignore this message because cache is full.
+                    this.logger.LogTrace("(-)[CACHE_IS_FULL]");
+                    return;
+                }
+
+                // If queue is not empty, add to queue instead of calling CM.
+                if (this.cachedHeaders.Count != 0)
+                {
+                    this.cachedHeaders.AddRange(headers);
+
+                    this.logger.LogTrace("{0} headers were added to cache, new cache size is {1}.", headers.Count, this.cachedHeaders.Count);
+                    this.logger.LogTrace("(-)[CACHED]");
+                    return;
+                }
+
+                var consensusFactory = new PosConsensusFactory();
+
+                ConnectNewHeadersResult result = await this.PresentHeadersLockedAsync(headers).ConfigureAwait(false);
```

Comment:
The solution doesn't build `Severity	Code	Description	Project	File	Line	Suppression State
Error	CS1503	Argument 1: cannot convert from 'System.Collections.Generic.List<NBitcoin.ProvenBlockHeader>' to 'System.Collections.Generic.List<NBitcoin.BlockHeader>'	Stratis.Bitcoin	C:\Users\user\Desktop\FNMainRepo\src\Stratis.Bitcoin\Consensus\PHConsensusManagerBehavior.cs	187	Active
`